### PR TITLE
Rewrite learning uploader

### PR DESCRIPTION
**Pull request by frankharkins**
The requirements for the learning uploader have changed since it was first created.

It was originally intended to be used locally by writers to quickly push content to the platform. I wrote it in Python so writers would not need to install node, and using string manipulations to interact with the REST API was good enough for just uploading the zip and linking the new file. It also included functionality for a writer to log into the database through a CLI.

Now, the uploader is only used in CI, so we don't care about logging in or sticking to Python, and we want to be able to do more complex manipulations such as adding new lessons and changing metadata. For this reason, this PR rewrites the uploader using the official [Directus SDK](https://docs.directus.io/guides/sdk/), which makes interacting with the API a lot easier.

### CODE REVIEW

File: scripts/tutorial-uploader/lib/api.test.ts
Comment (Eric-Arellano) [inline scripts/tutorial-uploader/lib/api.test.ts:27]: What if instead, you have a single variable inside the `describe` like `MOCK_TUTORIAL`? You can declare the object `as const` to ensure that one test doesn't overwrite the object and mess up other tests. Rather than `afterAll`, you'd use `afterEach` to remove `MOCK_TUTORIAL.slug`.

I think this might make things more clear. Right now, it's not very idiomatic to have `createdSlugs` and `generateTutorialData` outside the `desribe()`, and it took me a minute to convince myself that it's safe to use `afterAll` rather than `afterEach` because the tutorial slug will be unique.

--

If you take this recommendation, you don't need to use a random test ID. You can call it something like `test-tutorial`. 

Note that cleanup does not always happen properly, like if someone uses ctrl-c to exit out of the script. So, you might want to even delete this test entry in `beforeAll`.

File: scripts/tutorial-uploader/lib/api.test.ts
Comment (Eric-Arellano) [inline scripts/tutorial-uploader/lib/api.test.ts:46]:
```suggestion
  throw new Error("Tried to run tests against production! Set the env var LEARNING_API_URL to either staging or local (see tutorial-uploader/README.md)");
```

File: scripts/tutorial-uploader/lib/api.ts
Comment (Eric-Arellano) [inline scripts/tutorial-uploader/lib/api.ts:42]:
```suggestion
  readonly client: RestClient<LearningApiSchema>;
```
```suggestion
  client: RestClient<LearningApiSchema>;
```

File: scripts/tutorial-uploader/lib/api.ts
Comment (Eric-Arellano) [inline scripts/tutorial-uploader/lib/api.ts:111]: No need to handle multiple yet, but please do throw an exception so we know we need to deal with it

File: scripts/tutorial-uploader/lib/api.ts
Comment (Eric-Arellano) [inline scripts/tutorial-uploader/lib/api.ts:164]: Probably more clear to use `cd` from zx: https://google.github.io/zx/api#cd

Should this be changing the cwd for all future processes, or only this specific one?

File: scripts/tutorial-uploader/lib/local-tutorial-data.ts
Comment (Eric-Arellano) [inline scripts/tutorial-uploader/lib/local-tutorial-data.ts:50]: Lodash has functions like https://lodash.com/docs/#isString

**Review by Eric-Arellano — COMMENTED** [review]
This is exciting!

File: scripts/tutorial-uploader/lib/schema.ts
Comment (frankharkins) [reply scripts/tutorial-uploader/lib/schema.ts:17]: It's [SO copy/paste](https://stackoverflow.com/a/52331580) (I should have attributed; I'll leave a comment). Given an array `U[]`, this returns the type of each element (i.e. `U`).

File: scripts/tutorial-uploader/lib/schema.ts
Comment (frankharkins) [reply scripts/tutorial-uploader/lib/schema.ts:24]: Yep, this "collection" is a set of one-to-many mappings of "tutorials" to "tutorials_topics". Name is unfortunate but is out of our control. I'll add a comment.

File: scripts/tutorial-uploader/lib/schema.ts
Comment (frankharkins) [reply scripts/tutorial-uploader/lib/schema.ts:34]: It's an array of numbers. The API can either return a list of IDs or the object itself depending on the request you send. Here's the docs on setting up a schema: https://docs.directus.io/guides/sdk/types.html#adding-relational-fields.

File: scripts/tutorial-uploader/lib/schema.ts
Comment (frankharkins) [reply scripts/tutorial-uploader/lib/schema.ts:59]: Yeah it can return a uuid or the object itself; see https://github.com/Qiskit/documentation/pull/1421/files#r1618651601.

File: scripts/tutorial-uploader/lib/schema.ts
Comment (frankharkins) [reply scripts/tutorial-uploader/lib/schema.ts:17]: Actually, re-reading that post, it looks like we can use

```ts
type Unpacked<T> = T extends (infer U)[]
```

Which is in the [TS documentation](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-8.html).

File: scripts/tutorial-uploader/lib/api.test.ts
Comment (frankharkins) [reply scripts/tutorial-uploader/lib/api.test.ts:23]: That's better, thanks!

File: scripts/tutorial-uploader/lib/api.ts
Comment (arnaucasau) [reply scripts/tutorial-uploader/lib/api.ts:164]: I think that the intention is only to avoid having the full directory structure inside the zip file, but correct me if I'm wrong Frank. 

FWIW, for that, we could also modify the working directory with the `env -C` command in front of the `zip` command, something like `env -C ${path} zip -qr ${zippedFilePath} .` I like the use of `cd` as well, so that's just an alternative ([example](https://superuser.com/a/1842149))

File: scripts/tutorial-uploader/lib/schema.ts
Comment (frankharkins) [reply scripts/tutorial-uploader/lib/schema.ts:17]: FYI I actually came across a nice explanation of this the other day: https://youtu.be/eJ6R1knfsoc?feature=shared&t=1639

File: scripts/tutorial-uploader/lib/api.ts
Comment (frankharkins) [reply scripts/tutorial-uploader/lib/api.ts:164]: Yep, I don't think zx cd is ideal as I don't want to change the working directory for other commands. @Eric-Arellano @arnaucasau do you have a preference between `env -C` or `(cd ... && ... )`. I'll add an explanation comment either way.

Comment (Eric-Arellano) [reply scripts/tutorial-uploader/lib/api.ts:164]: `(cd ... && ... )` sgtm

Comment (arnaucasau) [reply scripts/tutorial-uploader/lib/api.ts:164]: Yeah, I agree with Eric :+1:

File: scripts/tutorial-uploader/lib/api.ts
Comment (frankharkins) [reply scripts/tutorial-uploader/lib/api.ts]: Fixed in 1696a8be75493290fafb945706a43689f6bfbc20; it now ignores non-English translations.

File: scripts/tutorial-uploader/lib/api.test.ts
Comment (frankharkins) [reply scripts/tutorial-uploader/lib/api.test.ts]: Ok nice, I assumed the jest tests ran in parallel so I added the IDs to avoid conflicts. Fixed in ad843ecec.

File: scripts/tutorial-uploader/lib/api.test.ts
Comment (Eric-Arellano) [inline scripts/tutorial-uploader/lib/api.test.ts:148]: Nice rewrite! This line is now repeated several times and could be a helper function

File: scripts/tutorial-uploader/README.md
Comment (frankharkins) [reply scripts/tutorial-uploader/README.md]: Added in a1095a15e0563fff4758bace546be59f77b868c4

File: scripts/tutorial-uploader/README.md
Comment (arnaucasau) [inline scripts/tutorial-uploader/README.md:27]: Nit:
```suggestion
  4. To test the script in this repo, export the following environment
```

File: scripts/tutorial-uploader/README.md
Comment (arnaucasau) [inline scripts/tutorial-uploader/README.md:12]: Nit: Maybe I'm wrong here and this is correct, but just in case :)

```suggestion
- Follow the instructions in the README in the saiba-api repo. Don't forget the
```

File: scripts/tutorial-uploader/lib/local-tutorial-data.ts
Comment (arnaucasau) [inline scripts/tutorial-uploader/lib/local-tutorial-data.ts:45]: I like how you are verifying if the object matches the interface with `verifyLocalTutorialData` giving a detailed error for the attributes that fail. 

Optional suggestion: I think we could make it more explicit that the function is also type-casting the objects into `LocalTutorialData` to make the code more readable and less prone to errors if in the future we need to change the verification somehow.

One option could be to change the function's name. Alternatively, we could also change the `verifyLocalTutorialData` function to return `void` and move the check before this return using a `forEach`. With this strategy the second map could to be `.map((i) => relativiseLocalPath(i as LocalTutorialData, path));`

File: tutorials/README.md
Comment (arnaucasau) [inline tutorials/README.md:25]: Nit:
```suggestion
make a new entry to `learning-api.conf.yaml`. The next time the deploy action is
```

**Review by arnaucasau — COMMENTED** [review]
Awesome work Frank! It looks great! :rocket:

File: scripts/tutorial-uploader/README.md
Comment (frankharkins) [reply scripts/tutorial-uploader/README.md:12]: No you're completely right, thanks!

File: scripts/tutorial-uploader/lib/local-tutorial-data.ts
Comment (frankharkins) [reply scripts/tutorial-uploader/lib/local-tutorial-data.ts]: Good idea, thanks! Done in ae35611a332d77b925b2d0d65ff43615a168afd5.

**Comment by frankharkins** [comment]
This is ready for final review, but it might be best to avoid merging until I get back in case there are teething issues.

File: scripts/tutorial-uploader/lib/api.test.ts
Comment (Eric-Arellano) [inline scripts/tutorial-uploader/lib/api.test.ts:35]: Given Abdón's request, should we no longer be mentioning staging? Maybe we should error if it's anything other than localhost? At that point, should we hardcode the value?

File: scripts/tutorial-uploader/lib/api.ts
Comment (Eric-Arellano) [inline scripts/tutorial-uploader/lib/api.ts:203]: What does this do? I'm surprised we'd want the title to have the random hex code and `tutorial.zip`

File: scripts/tutorial-uploader/lib/api.ts
Comment (Eric-Arellano) [inline scripts/tutorial-uploader/lib/api.ts:218]:
```suggestion
  ): Promise<void> {
```

File: scripts/tutorial-uploader/lib/api.ts
Comment (Eric-Arellano) [inline scripts/tutorial-uploader/lib/api.ts:245]: Idea prompted by this usage note: make most the functions private by putting `#` in front of the name so that it's clear what you expect other files to use or not. That reduces the risk of someone encountering this "gotcha" that they need to also use updateExistingTutorial

File: scripts/tutorial-uploader/lib/api.ts
Comment (Eric-Arellano) [inline scripts/tutorial-uploader/lib/api.ts:278]:
```suggestion
  async upsertTutorial(localData: LocalTutorialData): Promise<void> {
```

File: scripts/tutorial-uploader/lib/api.ts
Comment (Eric-Arellano) [inline scripts/tutorial-uploader/lib/api.ts:289]:
```suggestion
  async deleteTutorial(tutorialSlug: string): Promise<void> {
```

File: scripts/tutorial-uploader/lib/local-tutorial-data.ts
Comment (Eric-Arellano) [inline scripts/tutorial-uploader/lib/local-tutorial-data.ts:44]: Nice! This is definitely an improvement.

File: scripts/tutorial-uploader/lib/local-tutorial-data.ts
Comment (Eric-Arellano) [inline scripts/tutorial-uploader/lib/local-tutorial-data.ts:45]: Is the `as` still required, thanks to the type guard from assertIsLocalTutorialData?

```suggestion
  return parsed.map((i) => relativiseLocalPath(i, path));
```

File: scripts/tutorial-uploader/setup-for-testing.ts
Comment (Eric-Arellano) [inline scripts/tutorial-uploader/setup-for-testing.ts:32]: What's the impact of these becoming stale? If we use a new value in the config file that is not yet defined here, we'd fail in the test, which would be a signal to update this file?

File: tutorials/README.md
Comment (Eric-Arellano) [inline tutorials/README.md:37]:
```suggestion
You should always start with deploying to "Learning platform (staging)", and
check that your tutorial renders properly. Ask a teammate for the staging link
if you need it.
```

File: tutorials/learning-api.conf.yaml
Comment (Eric-Arellano) [inline tutorials/learning-api.conf.yaml:16]: Maybe add a note in this file that when you make changes to these two, update the setup test script.

**Review by Eric-Arellano — COMMENTED** [review]
This is an epic improvement!

File: scripts/tutorial-uploader/lib/api.ts
Comment (frankharkins) [reply scripts/tutorial-uploader/lib/api.ts:203]: From what I can tell, it's the human-readable title used in the CMS. Since the file gets destroyed after it's linked to the tutorial, it shouldn't matter much. If a human did see the file, I think the hex code helps convey it's for machines.

File: scripts/tutorial-uploader/lib/api.test.ts
Comment (frankharkins) [reply scripts/tutorial-uploader/lib/api.test.ts:35]: I think mentioning staging is ok, it's just the URLs that are the problem? I'll seek more guidance when we're back.

File: scripts/tutorial-uploader/lib/local-tutorial-data.ts
Comment (frankharkins) [reply scripts/tutorial-uploader/lib/local-tutorial-data.ts:45]: I tried that but unfortunately the type checker isn't that clever. It did work when I made a function that asserted `parsed` was of type `LocalTutorialData[]`, but I thought this was just as clear and a little simpler.

File: scripts/tutorial-uploader/setup-for-testing.ts
Comment (frankharkins) [reply scripts/tutorial-uploader/setup-for-testing.ts:32]: Yes, adding a category that doesn't exist would break the uploader. I'll make a check that verifies the list against production.

**Comment by frankharkins** [comment]
This is ready for final review, but we can't merge until the token has the right permissions.

File: scripts/tutorial-uploader/lib/api.test.ts
Comment (arnaucasau) [inline scripts/tutorial-uploader/lib/api.test.ts:51]: With this `readItem`, are you trying to read nested fields? If that's the case, FYI (probably you already know), I found this alternative [syntaxis](https://docs.directus.io/guides/sdk/getting-started.html#read-nested-fields). Maybe this could help with the `@ts-ignore` :thinking:. If that doesn't help or the `readItem` had another purpose, I still like how you have made it work. I didn't know about `@ts-ignore` before :rocket:

**Review by arnaucasau — APPROVED** [review]
Awesome work Frank :rocket:. I've been looking at the improvements you did from the comments Eric left last week and the code itself and all looks great! I left some feedback to try to understand better some parts and a possible nit

File: scripts/tutorial-uploader/lib/api.test.ts
Comment (frankharkins) [reply scripts/tutorial-uploader/lib/api.test.ts:51]: Awesome! Yes always best to fix the type rather than ignore when possible. That fixed all but one.

**Comment by frankharkins** [comment]
This comment is a reminder to update `learning-api.conf.yaml` before merging as tutorial data has changed since this PR was created.

**Comment by frankharkins** [comment]
This is ready to go! I'll use it deploy the QAOA tutorial then I'll move it to inner-source to hand off to the new team. Thanks for your patience.

File: scripts/tutorial-uploader/lib/api.test.ts
Comment (Eric-Arellano) [inline scripts/tutorial-uploader/lib/api.test.ts:145]: Nit that the convention is for this and `afterEach` to be near the top of the describe block

File: tutorials/learning-api.conf.yaml
Comment (Eric-Arellano) [inline tutorials/learning-api.conf.yaml:24]: This comment will need to change when migrating

File: tutorials/learning-api.conf.yaml
Comment (Eric-Arellano) [inline tutorials/learning-api.conf.yaml:4]: This is no longer accurate. I think this should be the learning team

**Review by Eric-Arellano — APPROVED** [review]
Great job!

**Pull request merged**
